### DISCUSSION TITLE
[QUESTION]: Can .NET interactive be embedded in a Web Application?

### DISCUSSION THREAD

**Discussion by rrekapalli** [dotnet/interactive#2612] [Q&A]
Hi there,

Does this library be embedded in a web application developed in ASP.NET/Angular/React? 

Thanks

**Comment by jonsequitur** [comment]
Yes. There are a number of packages on NuGet and you can see an embedding example in [this sample](https://github.com/dotnet/interactive/tree/main/samples/connect-wpf).

**Comment by colombod** [comment]
We are working on exposing the tools to configure your aspnet host with .NET Interactive in PR #802

**Comment by rrekapalli** [comment]
Thank you!

**Comment by PawelStadnicki** [comment]
@colombod aspnet host PR mentioned above seems to be merged now however there is no sample yet/blog/tutorial about it?
Hosting dotnet-interactive on custom webpage is highly anticipated IMHO

**Comment by colombod** [comment]
> @colombod aspnet host PR mentioned above seems to be merged now however there is no sample yet/blog/tutorial about it?
> Hosting dotnet-interactive on custom webpage is highly anticipated IMHO

Consuming .NET Interactive from a web page is not hard to begin but crafting a good experience takes time. We do not have yet documentation or tutorial for it but you can take a peak at a simple demo I have built [here](https://github.com/colombod/dotnet-interactive-experiments/tree/main/dotnet-interactive-html).

That is not official sample and would love your feedback, maybe you could help building a simple demo?

**Comment by PawelStadnicki** [comment]
Thanks, I will try it.
Will it be supported to use other kernels too ? I guess that GitHub Codespaces has the same functionality as VS Code. 
Wonder how much effort will be required to host on custom page same functionality as VS Code or GitHub Codespaces ( I assume that VS Code & Github Codespaces uses internal "notebook" support that must be adjusted for own purposes).

I can also guess that definitely there will be some blazor-flavour at some point.
I also know that MS internally uses some "trydotnetjs" that can also be public at some point
but grasping it all is hard without the full roadmap.

**Comment by colombod** [comment]
> Thanks, I will try it.
> Will it be supported to use other kernels too ? I guess that GitHub Codespaces has the same functionality as VS Code.
> Wonder how much effort will be required to host on custom page same functionality as VS Code or GitHub Codespaces ( I assume that VS Code & Github Codespaces uses internal "notebook" support that must be adjusted for own purposes).
> 
> I can also guess that definitely there will be some blazor-flavour at some point.
> I also know that MS internally uses some "trydotnetjs" that can also be public at some point
> but grasping it all is hard without the full roadmap.

You can already use any other kernel for .NET Interactive. That app has the same engine as the VSCode, Jupiter, and Jupiterlab integration. You can access the sharp, fsharp, powershell and js languages. To have same functionality as the code spaces you might want to look at the [nteract SDK](https://github.com/nteract/nteract) to create your frontend. Wiring up [Monaco editor](https://github.com/microsoft/monaco-editor) could be complex but I would not be surprised if the nteract ecosystem already has a Monaco integration

**Comment by GeorgeS2019** [comment]
Is it feasible to make .NET interactive hostable within Blazor Webassembly so we could make .NET interactive sharing [e.g. TorchSharp data science tutorials/examples](https://github.com/dotnet/TorchSharpExamples/issues/5#issue-1036760302) available through [Github Pages](https://swimburger.net/blog/dotnet/how-to-deploy-aspnet-blazor-webassembly-to-github-pages)

**Comment by jonsequitur** [comment]
A WASM-based kernel is feasible and is something we plan to investigate. Try .NET uses WASM today, so we'll look into building one or more new subkernels that can produce and run WASM along similar lines.

**Comment by AlexHedley** [comment]
Is this linked in anyway to `trydotnetjs`?
- https://github.com/dotnet/try/issues/95
- https://github.com/dotnet/try/blob/main/DotNetTry.md
- https://github.com/dotnet/try-samples#basics

**Reply by jonsequitur** [reply]
The Try .NET project has been merged into .NET Interactive but an updated version hasn't been released yet. https://github.com/dotnet/interactive/blob/main/docs/FAQ.md#how-is-net-interactive-related-to-try-net

**Comment by evertoncolling** [comment]
@jonsequitur any news on a WASM-based kernel (https://github.com/dotnet/interactive/issues/800#issuecomment-952373285)? I've been using Pyodide kernel in Jupyterlite for a while now and it works great. It would be nice to be able to run C# or F# code entirely on the browser.

**Reply by haslingerm** [reply]
I second this. It would allow me to provide interactive learning material for C# without requiring students to setup and get used to the Jupyter environment. Lots of potential there.

**Reply by jonsequitur** [reply]
The building blocks for this kind of experience are currently rolling out in the updated Try .NET service. There are some important caveats though:

* A service backend is still required to perform compilation, including servicing requests for language services such as completions and diagnostics.
* Dynamically adding NuGet packages is currently not supported.

The source code for the service can be found here: https://github.com/dotnet/try

**Reply by haslingerm** [reply]
@jonsequitur yes, I've seen that. For example the site https://dotnet.microsoft.com/en-us/learn/dotnet/in-browser-tutorial/1 for me just shows 'Compiling in the cloud...' forever (closed it after a few minutes) and never prints 'Hello World'.
Sure, that is just preview (I think?), but if MS themselves can't get it to work reliably we still have a way to go ;)

Not totally sure why we need an external service, though - doesn't Mono (used for Blazor) have an interpreter? Should be good enough for some small snippets? I did not look through the actual code, so this is just an uneducated guess.
Requiring a backend service is a huge deal, because it requires a sufficiently sandboxed environment to execute raw user input. I would very much prefer to only run in the user's browser, so that the worst that can happen is them killing their own browser. 

For NuGets I don't think dynamically adding them is really necessary, if we can define a set when creating the page (or whatever deployment unit we will have).

Thanks for your continued work on this feature! It will help a lot to show how flexible modern C#/.NET is these days.

**Reply by AlexHedley** [reply]
One option I was working on was using Blazor [dotnet-interactive-blazor-wasm](https://github.com/alex-hedley/dotnet-interactive-blazor-wasm) to host the backend and call to that, after seeing a video from Steve S https://www.youtube.com/watch?v=gKX-cdqnb8I, but it’s very much WIP.

**Reply by jonsequitur** [reply]
> Not totally sure why we need an external service, though - doesn't Mono (used for Blazor) have an interpreter? Should be good enough for some small snippets? 

Compilation happens in the cloud so that the entire compilation toolchain and resources such as XML documentation files don't need to be downloaded to the browser. This provides a very significant performance improvement over trying to do compilation in the browser. Once compiled, an assembly containing the user's code is executed in the browser using Blazor WASM.

You can see a breakdown of the call flows here: https://github.com/dotnet/interactive/tree/main/src/Microsoft.DotNet.Interactive.CSharpProject#try-net--microsoft-learn-flow

> Requiring a backend service is a huge deal, because it requires a sufficiently sandboxed environment to execute raw user input. 

A sandbox isn't needed because the user's code is executed within their own browser.

**Reply by haslingerm** [reply]
> A sandbox isn't needed because the user's code is executed within their own browser.

As long as the users can't add arbitrary source generators/analyzers or do other shenanigans. But I guess in this scenario the environment can be well controlled, so that shouldn't be an issue, true.

I run a service where students can enter code and have it checked—not only if it compiles, but also if a set of unit tests passes. So, I do have user code execution on the server and needed a sandbox for that, which is why my thoughts drifted in that direction.

**Reply by AlexHedley** [reply]
I built [BlazorInteractive](https://alex-hedley.github.io/BlazorInteractive/) in WASM to run code snippets, a bit like [.NET Fiddle](http://dotnetfiddle.net/) but it meant I could host it on GitHub Pages and not worry about a server, but it was a lot of work and still needs a load of features. If I could just use **Interactive** instead I’d get all the power for free.

**Reply by hawkerm** [reply]
I feel like this is an assembly loading thing where .NET Interactive is assuming it can load assemblies from disk. In Alex's case he already loaded the assemblies from the Blazor framework output behind the scenes to use CodeAnalysis.CSharp. For general .NET and WASM you can load them manually as well as part of the output, this is what's done in this Uno demo here: https://github.com/jeromelaban/uno.skiasharpfiddle. 

Seems like it'd be a matter of providing an option for loading assemblies from a general location vs. having .NET interactive take on that responsibility and assume it knows where they are?

**Comment by PawelStadnicki** [comment]
I was working on sth like this: hosting Interactive on the Blazor app that runs F# code on a client browser, using my own mini-editor based on Monaco.
All the code was created with F# (even JS with fable compiler), and Blazor is only to serve the wasm purpose.
Currently, it works ok on the server-blazor but the client mode (wasm) fails because of the F# Kernel limitations (I believe the same will be for C#).

F# kernel uses standard F# script engine (fsi) which is heavily Input-Output related (console logging, files etc), a lot of stuff that will not work on wasm or must be tailored/marked for that purpose. I asked the F# team to make the FSI lighter without any unnecessary computation tied to the environment. 

I expect it will not be a part of the .NET 9, though, but when I have time I will try to do my experiments

Worth mentioning is also the fact, that with .NET 9 (tentative) wasm is going to support real multithreading :)

**Reply by AlexHedley** [reply]
I’ve got an F# WIP branch on mine but it’s a bit more involved, as you say.

**Reply by jonsequitur** [reply]
The C# Script engine is quite different from FSI and can potentially run under WASM but in both cases, there will be significant performance problems due to the number of files that are needed in the browser, especially for language services.

**Reply by PawelStadnicki** [reply]
I somehow missed the info that multithreading in Blazor-wasm is officially postponed till .NET 10 ...
I don't see the point for my use cases to run the code on the client that cannot be multi-threaded, taking into account another valid performance point from @jonsequitur.

So much work is needed in both Wasm and F# within at least a 1.5-year horizon

**Comment by GeorgeS2019** [comment]
Any update, embedded in Blazor in 2025?
